File: acp-sdk-go/overlay_framework_test.go
package acp

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"path/filepath"
	"testing"
)

func TestOverlayRuntimeExposesWellKnownHeadersAndRejectsNonObjectBody(t *testing.T) {
	tempDir := t.TempDir()
	agent, err := LoadOrCreate("agent:overlay.runtime@localhost:9541", &AcpAgentOptions{
		StorageDir:        filepath.Join(tempDir, "runtime-agent"),
		Endpoint:          "http://localhost:9541/acp/inbox",
		AllowInsecureHTTP: true,
		DiscoveryScheme:   "http",
	})
	if err != nil {
		t.Fatalf("runtime agent: %v", err)
	}
	runtime, err := NewOverlayFrameworkRuntime(agent, "http://localhost:9541", func(payload map[string]any) map[string]any {
		return map[string]any{
			"accepted": true,
			"echo":     payload,
		}
	}, nil)
	if err != nil {
		t.Fatalf("overlay runtime: %v", err)
	}
	headers := OverlayWellKnownHeaders()
	if headers["Cache-Control"] != "public, max-age=300" {
		t.Fatalf("unexpected cache-control header")
	}
	wellKnown, err := runtime.WellKnownDocument()
	if err != nil {
		t.Fatalf("well-known document should be built: %v", err)
	}
	if wellKnown["agent_id"] != "agent:overlay.runtime@localhost:9541" {
		t.Fatalf("unexpected well-known agent_id")
	}
	if wellKnown["version"] != "1.0" {
		t.Fatalf("unexpected well-known version")
	}
	invalidResponse := runtime.HandleMessageBody([]any{"invalid"})
	if invalidResponse.StatusCode != 400 {
		t.Fatalf("invalid request should return 400")
	}
	if invalidResponse.Body["state"] != "FAILED" {
		t.Fatalf("invalid request should be FAILED")
	}
	if invalidResponse.Body["reason_code"] != "POLICY_REJECTED" {
		t.Fatalf("invalid request reason code mismatch")
	}
}

func TestOverlayClientBootstrapsWellKnownAndSendsPayload(t *testing.T) {
	tempDir := t.TempDir()
	requestCounts := map[string]int{}
	var wellKnown map[string]any
	var identityPayload map[string]any
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		requestCounts[request.Method+" "+request.URL.Path]++
		switch {
		case request.Method == http.MethodGet && request.URL.Path == "/.well-known/acp":
			writer.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(writer).Encode(wellKnown)
		case request.Method == http.MethodGet && request.URL.Path == "/api/v1/acp/identity":
			writer.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(writer).Encode(identityPayload)
		case request.Method == http.MethodPost && request.URL.Path == "/acp/inbox":
			writer.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(writer).Encode(map[string]any{"status": "accepted"})
		default:
			http.NotFound(writer, request)
		}
	}))
	defer server.Close()

	receiver, err := LoadOrCreate("agent:overlay.receiver@localhost:9542", &AcpAgentOptions{
		StorageDir:        filepath.Join(tempDir, "receiver"),
		Endpoint:          server.URL + "/acp/inbox",
		AllowInsecureHTTP: true,
		DiscoveryScheme:   "http",
	})
	if err != nil {
		t.Fatalf("receiver agent: %v", err)
	}
	wellKnown, err = receiver.BuildWellKnownDocument(server.URL, "")
	if err != nil {
		t.Fatalf("receiver well-known: %v", err)
	}
	identityPayload = map[string]any{
		"identity_document": receiver.IdentityDocument,
	}

	sender, err := LoadOrCreate("agent:overlay.sender@localhost:9543", &AcpAgentOptions{
		StorageDir:        filepath.Join(tempDir, "sender"),
		AllowInsecureHTTP: true,
		DiscoveryScheme:   "http",
	})
	if err != nil {
		t.Fatalf("sender agent: %v", err)
	}
	client := NewOverlayClient(sender)
	response, err := client.SendACP(
		server.URL,
		map[string]any{"kind": "overlay-go-test"},
		"",
		"overlay:go:outbound",
		DeliveryAuto,
		120,
	)
	if err != nil {
		t.Fatalf("overlay outbound send: %v", err)
	}

	target, ok := response["target"].(*OverlayTarget)
	if !ok || target == nil {
		t.Fatalf("target should be present")
	}
	if target.AgentID != "agent:overlay.receiver@localhost:9542" {
		t.Fatalf("target agent mismatch")
	}
	if target.WellKnownURL != server.URL+"/.well-known/acp" {
		t.Fatalf("well-known URL mismatch")
	}
	sendResult, ok := response["send_result"].(SendResult)
	if !ok {
		t.Fatalf("send_result should be present")
	}
	if len(sendResult.Outcomes) != 1 {
		t.Fatalf("expected one delivery outcome")
	}
	state := sendResult.Outcomes[0].State
	if state != StateDelivered && state != StateAcknowledged {
		t.Fatalf("unexpected delivery state: %s", state)
	}
	if requestCounts["GET /.well-known/acp"] == 0 {
		t.Fatalf("well-known endpoint should be called")
	}
	if requestCounts["GET /api/v1/acp/identity"] == 0 {
		t.Fatalf("identity endpoint should be called")
	}
	if requestCounts["POST /acp/inbox"] == 0 {
		t.Fatalf("inbox endpoint should be called")
	}
}

File: acp-sdk-typescript/tests/overlay-framework.test.ts
import { createServer, IncomingMessage, ServerResponse } from "node:http";
import { AddressInfo } from "node:net";
import { mkdtempSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { describe, expect, it } from "vitest";
import { AcpAgent } from "../src/agent.js";
import { verifyIdentityDocument } from "../src/identity.js";
import { defaultAgentOptions } from "../src/options.js";
import { OverlayClient, OverlayFrameworkRuntime } from "../src/overlayFramework.js";
import { JsonMap, JsonValue } from "../src/jsonSupport.js";

function readBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    req.on("data", (chunk) => chunks.push(Buffer.from(chunk)));
    req.on("end", () => resolve(Buffer.concat(chunks).toString("utf-8")));
    req.on("error", reject);
  });
}

function writeJson(res: ServerResponse, payload: JsonValue, status = 200): void {
  const body = JSON.stringify(payload);
  res.statusCode = status;
  res.setHeader("content-type", "application/json");
  res.end(body);
}

describe("Overlay framework runtime", () => {
  it("exposes well-known headers and rejects invalid body", async () => {
    const root = mkdtempSync(join(tmpdir(), "acp-ts-overlay-runtime-"));
    try {
      const receiver = await AcpAgent.loadOrCreate("agent:receiver.framework@localhost:9551", {
        ...defaultAgentOptions(),
        storage_dir: join(root, "receiver"),
        endpoint: "http://localhost:9551/acp/inbox",
        allow_insecure_http: true,
        discovery_scheme: "http"
      });
      const runtime = OverlayFrameworkRuntime.create(
        receiver,
        "http://localhost:9551",
        (payload) => ({ accepted: true, echo: payload })
      );
      const headers = OverlayFrameworkRuntime.wellKnownHeaders();
      expect(headers["Cache-Control"]).toBe("public, max-age=300");
      const wellKnown = runtime.wellKnownDocument();
      expect(wellKnown.agent_id).toBe("agent:receiver.framework@localhost:9551");

      const invalidResponse = await runtime.handleMessageBody(["invalid"]);
      expect(invalidResponse.status_code).toBe(400);
      expect(invalidResponse.body.state).toBe("FAILED");
      expect(invalidResponse.body.reason_code).toBe("POLICY_REJECTED");
    } finally {
      rmSync(root, { recursive: true, force: true });
    }
  });

  it("overlay client bootstraps from well-known and sends payload", async () => {
    const root = mkdtempSync(join(tmpdir(), "acp-ts-overlay-client-"));
    const receiver = await AcpAgent.loadOrCreate("agent:receiver.framework@localhost:9552", {
      ...defaultAgentOptions(),
      storage_dir: join(root, "receiver"),
      endpoint: "http://localhost:9552/acp/inbox",
      allow_insecure_http: true,
      discovery_scheme: "http"
    });
    expect(verifyIdentityDocument(receiver.identity_document)).toBe(true);
    const server = createServer(async (req, res) => {
      if (req.method === "GET" && req.url === "/.well-known/acp") {
        const port = (server.address() as AddressInfo).port;
        const baseUrl = `http://127.0.0.1:${port}`;
        writeJson(res, {
          agent_id: receiver.agentId(),
          identity_document: `${baseUrl}/api/v1/acp/identity`,
          transports: {
            http: {
              endpoint: `${baseUrl}/acp/inbox`
            }
          },
          version: "1.0",
          security_profile: "http"
        });
        return;
      }
      if (req.method === "GET" && req.url === "/api/v1/acp/identity") {
        writeJson(res, { identity_document: receiver.identity_document });
        return;
      }
      if (req.method === "POST" && req.url === "/acp/inbox") {
        await readBody(req);
        writeJson(res, { status: "accepted" });
        return;
      }
      writeJson(res, { error: "not found" }, 404);
    });
    await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", () => resolve()));

    try {
      const port = (server.address() as AddressInfo).port;
      const baseUrl = `http://127.0.0.1:${port}`;
      const sender = await AcpAgent.loadOrCreate("agent:sender.framework@localhost:9553", {
        ...defaultAgentOptions(),
        storage_dir: join(root, "sender"),
        allow_insecure_http: true,
        discovery_scheme: "http"
      });
      const client = OverlayClient.create(sender);
      const response = await client.sendAcp(baseUrl, { kind: "runtime-outbound" } as JsonMap);
      const target = response.target as JsonMap;
      expect(target.agent_id).toBe("agent:receiver.framework@localhost:9552");
      expect(target.well_known_url).toBe(`${baseUrl}/.well-known/acp`);
      const sendResult = response.send_result as JsonMap;
      const outcomes = sendResult.outcomes as JsonValue[];
      expect(Array.isArray(outcomes)).toBe(true);
      expect(outcomes.length).toBe(1);
    } finally {
      await new Promise<void>((resolve) => server.close(() => resolve()));
      rmSync(root, { recursive: true, force: true });
    }
  });
});
